staging/src/k8s.io/apiserver/pkg/storage/etcd3/store.go

```diff
@@ -738,28 +747,56 @@ func (s *store) getStateFromObject(obj runtime.Object) (*objState, error) {
 	return state, nil
 }
 
-func (s *store) updateState(st *objState, userUpdate storage.UpdateFunc) (runtime.Object, uint64, error) {
-	ret, ttlPtr, err := userUpdate(st.obj, *st.meta)
+var errStaleTTL = errors.New("updateState needs current objState for TTL calculation")
+
+func (s *store) updateState(st *objState, userUpdate storage.UpdateFunc, mustCheckData bool) (runtime.Object, uint64, error) {
+	ret, ttlPtr, err := userUpdate(st.obj, st.meta)
 	if err != nil {
 		return nil, 0, err
 	}
 
-	if err := s.versioner.PrepareObjectForStorage(ret); err != nil {
-		return nil, 0, fmt.Errorf("PrepareObjectForStorage failed: %v", err)
-	}
 	var ttl uint64
 	if ttlPtr != nil {
+		// if userUpdate asserts a TTL and we are using a cached object, it means that we passed a zero TTL
+		// to userUpdate as the existing object's TTL (because the cache does not know the TTL value).
+		// send a specific error in this case to signal that we need to get the actual object from ETCD
+		// so we can extract the lease's current TTL.  this check assumes that userUpdate will only assert
+		// a non-nil TTL when the associated object actually needs a TTL (or wants to remove a TTL).
+		// registry.Store.Update has the "correct" behavior in this regard.
+		if mustCheckData {
+			return nil, 0, errStaleTTL
+		}
 		ttl = *ttlPtr
 	}
+	// when ttlPtr is nil it would be nice to be able to use the value stored in st.meta.TTL
+	// as the TTL.  but the only way for us to know if that value is correct when mustCheckData
+	// is true is to fetch the latest object.  since ttlPtr is almost always nil, we would lose
+	// all benefit from using the suggestion object if we attempted a fetch.  we also cannot
+	// blindly use st.meta.TTL because it is zero when the state of the object is restored
+	// from the cache and thus would lead to inconsistent behavior (i.e. sometimes we keep the correct
+	// TTL for the object and other times we set the TTL to zero so the object does not expire).
+	// thus for the sake of consistency we assume a nil TTL to mean zero.
+
+	if err := s.versioner.PrepareObjectForStorage(ret); err != nil {
+		return nil, 0, fmt.Errorf("PrepareObjectForStorage failed: %v", err)
+	}
+
 	return ret, ttl, nil
 }
 
 // ttlOpts returns client options based on given ttl.
 // ttl: if ttl is non-zero, it will attach the key to a lease with ttl of roughly the same length
-func (s *store) ttlOpts(ctx context.Context, ttl int64) ([]clientv3.OpOption, error) {
+// if ttl is the same as st.meta.TTL, reuse the lease specified in st.lease
+func (s *store) ttlOpts(ctx context.Context, ttl int64, st *objState) ([]clientv3.OpOption, error) {
 	if ttl == 0 {
 		return nil, nil
 	}
+
+	// we want to keep the same TTL as before, so reuse the same lease
+	if st != nil && ttl == st.meta.TTL && st.lease != clientv3.NoLease {
+		return []clientv3.OpOption{clientv3.WithLease(st.lease)}, nil
+	}
+
 	id, err := s.leaseManager.GetLease(ctx, ttl)
```

Comment:
Is the difference between `s.leaseManager.GetLease(ctx, ttl)` and `ttl == st.meta.TTL` is that the lease manager call would create options with a remaining TTL == to the provided ttl, but `ttl == st.meta.TTL` will return the lease ID previously used and not extend the remaining TTL?  I.e. for `st.meta.TTL = 5` where the lease exists and has a remaining TTL of 3 seconds, because 2 seconds have already passed `ttlOps(ctx, 5, st)` would return a ID of existing TTL, with 3 seconds remaining, but for another `objState.meta.TTL = 30` where lease exists, `ttlOps(ctx, 5, st)` returns a lease with remaining TTL of 5 seconds?

Reply:
> Is the difference between `s.leaseManager.GetLease(ctx, ttl)` and `ttl == st.meta.TTL` is that the lease manager call would create options with a remaining TTL == to the provided ttl, but `ttl == st.meta.TTL` will return the lease ID previously used and not extend the remaining TTL?

Yeah, without the new `if` statement I added, even when the same TTL is requested, a new lease would get created and attached to the key (and the old lease would be unattached from the key).  I assume the point of passing `update` and `existing` to the `TTLFunc` is to allow flows that basically say "do not change the TTL once it is set."  I doubt it is a super common flow, but it does not seem wise to keep creating new leases when the old one can be reused.

> I.e. for `st.meta.TTL = 5` where the lease exists and has a remaining TTL of 3 seconds, because 2 seconds have already passed `ttlOps(ctx, 5, st)` would return a ID of existing TTL, with 3 seconds remaining

`st.meta.TTL` is assigned to the _remaining_ TTL value (not the originally requested TTL), meaning in this example it would be `3`.  The `TTLFunc` tells you how much time is remaining via the `existing` parameter.

If it was based on the originally requested TTL, the bump case used by events would not work: you request 30s the first time, and if you request 30s again it would not bump and reuse the lease.

> but for another `objState.meta.TTL = 30` where lease exists, `ttlOps(ctx, 5, st)` returns a lease with remaining TTL of 5 seconds?

I believe this is correct.

Reply:
> The lease manager already ensures that leases are reused when appropriate.

It just does simple math based on last lease it created.  It cannot correctly reuse leases over time for the same key.  It shares leases across keys created at close intervals within some approximation of the requested TTL.

> One concrete example of how ttl == st.meta.TTL might be suprising is that if one intends to renew the lease to a desired TTL to X seconds, so they return that value, but meta.TTL just happens to be X so instead they get the remaining time on the existing lease, not a full X seconds.

In general the TTL value is fuzzy with the lease manager in play.  Realistically one should expected the TTL to be within +/- 30 seconds of the requested time.  You asked for some TTL (from "now") and we honored it approximately.

Reply:
> It just does simple math based on last lease it created. It cannot correctly reuse leases over time for the same key. It shares leases across keys created at close intervals within some approximation of the requested TTL.

We shouldn't need to care what etcd lease is used, only that the "at least as long as X" guarantee is adhered to.  The lease manager achieves this and does it in a way that offers some efficiency gains.

> In general the TTL value is fuzzy with the lease manager in play. Realistically one should expected the TTL to be within +/- 30 seconds of the requested time. You asked for some TTL (from "now") and we honored it approximately.

Anyone expecting +/- 30 seconds is in trouble, because that's not the guarantee. The guarantee provided by etcd is that the lease will expire no sooner than the TTL. Leader elections extend out TTL without any theoretical bounds and we recently merged https://github.com/etcd-io/etcd/pull/9924 to try to reduce the probability of this being an issue in practice.

The problem as I see it with this change is that if one returns 5 s as expected, they expect the lease to last at least 5 seconds, but if the st.meta.TTL and the existing lease only has 3 seconds left, then they got 3 seconds instead of the desired 5, which does not honor their intent.

Reply:
> The problem as I see it with this change is that if one returns 5 s as expected, they expect the lease to last at least 5 seconds, but if the st.meta.TTL and the existing lease only has 3 seconds left, then they got 3 seconds instead of the desired 5, which does not honor their intent.

I do not see how this would ever happen.  If you asked for 5 and 3 s was remaining, you would get 5.  If you said 3, you would get 3 and the lease would be reused.

Reply:
> Anyone expecting +/- 30 seconds is in trouble

I meant that the TTL requested from etcd would be within some range, not that etcd would honor it explicitly within that range.

Reply:
For now I have left this in (because it is easy to drop the `if` statement and update the tests to not require lease reuse).